src/CodeAnalysis/CSharp/MetaAnalyzers/CSharpImmutableObjectMethodAnalyzer.cs

```diff
@@ -0,0 +1,108 @@
+        {
+            get
+            {
+                return ImmutableArray.Create(DoNotIgnoreReturnValueDiagnosticRule);
+            }
+        }
+
+        private static readonly string s_solutionFullName = @"Microsoft.CodeAnalysis.Solution";
+        private static readonly string s_projectFullName = @"Microsoft.CodeAnalysis.Project";
+        private static readonly string s_documentFullName = @"Microsoft.CodeAnalysis.Document";
+        private static readonly string s_syntaxNodeFullName = @"Microsoft.CodeAnalysis.SyntaxNode";
+
+        // A list of known immutable object names
+        private static readonly List<string> s_immutableObjectNames = new List<string>()
+        {
+            s_solutionFullName,
+            s_projectFullName,
+            s_documentFullName,
+            s_syntaxNodeFullName,
+        };
+
+        private static readonly string s_Add = "Add";
+        private static readonly string s_Remove = "Remove";
+        private static readonly string s_Replace = "Replace";
+        private static readonly string s_With = "With";
+
+        private static readonly List<string> s_immutableMethodNames = new List<string>()
+        {
+            s_Add,
+            s_Remove,
+            s_Replace,
+            s_With,
+        };
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterCodeBlockAction(AnalyzeCodeBlockForIgnoredReturnValues);
+        }
+
+        public void AnalyzeCodeBlockForIgnoredReturnValues(CodeBlockAnalysisContext context)
+        {
+            var model = context.SemanticModel;
+            var syntaxNode = context.CodeBlock;
+
+            //Find invocations that ignore the return value
+            //We're looking for invocations that are not within an equals value clause or a return statement.
+            var candidateInvocations = syntaxNode.DescendantNodes()
+                .OfType<InvocationExpressionSyntax>()
+                .Where(n => !(n.Ancestors().Any(m => m is EqualsValueClauseSyntax || m is ReturnStatementSyntax)));
+
+            foreach (var candidateInvocation in candidateInvocations)
+            {
+                //If we can't find the method symbol, quit
+                var methodSymbol = model.GetSymbolInfo(candidateInvocation).Symbol as IMethodSymbol;
+                if (methodSymbol == null)
+                    continue;
+
+                //If the method doesn't start with something like "With" or "Replace", quit
+                string methodName = methodSymbol.Name;
+                if (!s_immutableMethodNames.Any(n => methodName.StartsWith(n)))
+                    continue;
+
+                //If we're not in one of the known immutable types, quit
+                var parentName = methodSymbol.ContainingType.ToString();
+                var baseTypesAndSelf = methodSymbol.ContainingType.GetBaseTypes().Select(n => n.ToString()).ToList();
```

Comment:
Instead of allocating a list in here I would just use a foreach and handle the parent type separately